### PR TITLE
[mlir][scf] Relax requirements for loops fusion

### PR DESCRIPTION
**Pull request by fabrizio-indirli**
Enable the fusion of parallel loops also when the 1st loop contains multiple write accesses to the same buffer, if the accesses are always on the same indices.
Fix LIT test cases whose loops were not being fused.

### CODE REVIEW

**Comment by fabrizio-indirli** [comment]
This patch attempts to enable the fusion of parallel loops in more cases, by relaxing some of the requirements:

### Allowing fusion when only one function argument is being written
Before this patch, the following loops would not be fused:
```
func.func @fuse_two(%A: memref<2x2xf32>, %B: memref<2x2xf32>,
                    %C: memref<2x2xf32>, %result: memref<2x2xf32>) {
  ...  // constants definitions
  %sum = memref.alloc()  : memref<2x2xf32>
  scf.parallel (%i, %j) = (%c0, %c0) to (%c2, %c2) step (%c1, %c1) {
    %B_elem = memref.load %B[%i, %j] : memref<2x2xf32>
    %C_elem = memref.load %C[%i, %j] : memref<2x2xf32>
    %sum_elem = arith.addf %B_elem, %C_elem : f32
    memref.store %sum_elem, %sum[%i, %j] : memref<2x2xf32>
    scf.reduce
  }
  scf.parallel (%i, %j) = (%c0, %c0) to (%c2, %c2) step (%c1, %c1) {
    %sum_elem = memref.load %sum[%i, %j] : memref<2x2xf32>
    %A_elem = memref.load %A[%i, %j] : memref<2x2xf32>
    %product_elem = arith.mulf %sum_elem, %A_elem : f32
    memref.store %product_elem, %result[%i, %j] : memref<2x2xf32>
    scf.reduce
  }
  memref.dealloc %sum : memref<2x2xf32>
  return
}
```
... because they were deemed as having possible aliasing issues, even though only one of the function arguments is being written.
[This test case is actually taken from the LIT suite that was already in place](https://github.com/llvm/llvm-project/blob/543cf08636f3a3bb55dddba2e8cad787601647ba/mlir/test/Dialect/SCF/parallel-loop-fusion.mlir#L27), but the failure wasn't being picked up due to a missing check in the test: thus, the loops weren't being fused, but the test passed anyway.
In my patch, I'm checking that at most one buffer from the function's arguments is being written among the two loops: if so, I assume there is no risk of aliasing. **Is this a correct assumption?**

### Allowing fusion when 1st loop contains multiple writes to a buffer that is then read, but always on the same indices
Before this patch, the following loops would not be fused:
```
  scf.parallel (%arg0) = (%c0) to (%c5) step (%c1) {
   ...
   memref.store %c2 %alloc[%arg0]
   ...
   memref.store %c2 %alloc[%arg0]
   scf.reduce
 }
 scf.parallel (%arg0) = (%c0) to (%c5) step (%c1) {
   ...
   %2 = memref.load %alloc[%arg0]
 }
```
... because the first loop contains multiple stores on the same buffer, that is then read by the 2nd loop.
However I believe that there should be no problem in fusing these loops when the 'store's access exactly the same indices, like in the example above.

**Comment by Hardcode84** [comment]
Thanks for the PR, I need some more time to think about this cases.

**Comment by fabrizio-indirli** [comment]
For the 1st case, it's been pointed out to me that there could be an issue with global variables: to exclude potential aliasings when only one function operand is being written, we would also need to make sure that the function does not access any globals to which the operand could be aliased. I'll add this check as well

**Comment by Hardcode84** [comment]
I think first case with multiple writes to the same indices is fine, 
but I believe I have a counter-example for the second case with func arguments:
```
func.func @fuse_two(%A: memref, %B: memref) {
    %temp = memref.alloc
    scf.parallel (%i, %j) = (%c0, %c0) to (%c2, %c2) step (%c1, %c1) {
        %A_elem = memref.load %A[%i, %j]
        memref.store %A_elem, %temp[%i, %j] : memref<2x2xf32>
    }
    scf.parallel (%i, %j) = (%c0, %c0) to (%c2, %c2) step (%c1, %c1) {
        %elem = memref.load %temp[%i, %j]
        memref.store %elem, %B[%i, %j] : memref<2x2xf32>
    }
}
```
If %A and %B alias same array with an different offset, the fusion will not be legal.
And generally, I would prefer to be conservative about handling functions arguments.
What you probably want is some sort of `restrict` attribute on function args, so user can 
guarantee args do not alias. Upstream alias analysis doesn't support this, but this is the
reason I did aliasing hook configurable.

Regarding broken tests, they probably can be fixed by making some memrefs local allocations instead
of functions args.

**Comment by fabrizio-indirli** [comment]
Thank you for having a look @Hardcode84,
I believe you're correct, I had missed those potential aliasing cases.
I have removed that part of the patch (the one handling the 2nd case) and I've modified the LIT tests.
Some of the test cases that I've made are a bit "artificial" (e.g. now the 1st loop reads from the output tensor to avoid the aliasing issue, but this is probably uncommon in real scenarios) but they should pass

File: mlir/test/Dialect/SCF/parallel-loop-fusion.mlir
Comment (Hardcode84) [inline mlir/test/Dialect/SCF/parallel-loop-fusion.mlir:482]: nit: newline at the end of file

File: mlir/test/Dialect/SCF/parallel-loop-fusion.mlir
Comment (Hardcode84) [inline mlir/test/Dialect/SCF/parallel-loop-fusion.mlir:342]: nit: while we are here, can you change all constant checks to `CHECK-DAG` as their exact order is not strictly guaranteed.

**Review by Hardcode84 — APPROVED** [review]

File: mlir/test/Dialect/SCF/parallel-loop-fusion.mlir
Comment (fabrizio-indirli) [reply mlir/test/Dialect/SCF/parallel-loop-fusion.mlir]: **✓** Done

**Comment by fabrizio-indirli** [comment]
Thanks for the review @Hardcode84 !
Unfortunately I don't have write access, do you think you could merge this patch please, if there are no further modifications needed to this change?

Note: I noticed that the CI check on Windowx x64 is failing, but the failure seems to be [unrelated to my change](https://buildkite.com/llvm-project/github-pull-requests/builds/34297), but rather to the CI env: CMake is returning an error "_Could NOT find Python3_"

**Comment by Hardcode84** [comment]
Merged, thanks for the contribution!

Windows bot is quite unreliable now, yes.

**Pull request merged**
